[null-aggr] Add null handling support in `mode` aggregation

gortiz opened the pull request:
This PR adds null handling support in the `mode` aggregation. Specifically, it modifies the code to ignore null values when `mode` is evaluated with `nullHandlingEnabled`, trying to follow the following postgres logic:

```psql
postgres=# create table myTable(myInt INT);
CREATE TABLE
postgres=# insert into myTable(myInt) VALUES (null), (1), (null);
INSERT 0 3
postgres=# SELECT mode() WITHIN GROUP (ORDER BY myInt) AS mode FROM myTable;
 mode 
------
    1
(1 row)

```

It is recommended to review this PR assuming PR #12226 is merged, which is equivalent to compare this PR from commit commit f966b1b295134a08a3b201dadcedd91ca7d01c0a to head. As you can see there, this PR also changes the test and shows that when null handling is enabled, nulls are ignored.

walterddr commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:
```diff
@@ -0,0 +1,285 @@
+import org.roaringbitmap.RoaringBitmap;
+
+
+public abstract class NullableSingleInputAggregationFunction<I, F extends Comparable>
```
iiuc this class would be extensible from all null-handling aggregate features? or do we plan to only use this for mode and not the other ones?

gortiz replied:
TL;DR: That class should be used by all aggregations once we migrate them.

I started adapting most (if not all) aggregation functions, but the PR was huge and I didn't write the test. So I decided to freeze that code. I pushed my progress to a personal branch in case I need to consult it (you can read it in https://github.com/gortiz/pinot/tree/null-aggr-bak) and I'm bringing changes from that branch into function specific branches.

gortiz replied:
I've moved this class to https://github.com/apache/pinot/pull/12215, which is going to be the base PR of this family of PRs

gortiz replied:
Finally I decided to move this code to https://github.com/apache/pinot/pull/12236, given #12215 is focused on the fluent framework itself. Anyway, both #12215 and #12236 should be merged (or discarded) soon.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:
```diff
@@ -0,0 +1,368 @@
+
+
+  @FunctionalInterface
+  public interface BatchConsumer {
```
I can see them simplify the code, but from past experience, the functional interface can significant impact the performance (at least for java 8). Aggregation function is one of the most performance sensitive code, and these function are applied on a per value basis. Can we do some micro benchmark to ensure there is no overhead?

gortiz replied:
I've created a benchmark in another PR (marked as draft). See https://github.com/apache/pinot/pull/12354

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:
```diff
@@ -0,0 +1,368 @@
+import org.roaringbitmap.RoaringBitmap;
+
+
+public abstract class NullableSingleInputAggregationFunction<I, F extends Comparable>
```
Are we going to make all aggregation able to take `null` values? If so, maybe consider adding the new methods directly to `BaseSingleInputAggregationFunction`

gortiz replied:
The plan is to eventually add it to all operations when it makes sense. In the medium term the idea is to support null handling in all operations that are not multi valued or not super pinot specific, so we can increase the compatibility with _standard_ SQL

gortiz commented:
@Jackie-Jiang this should be ready to merge. Can you take a look? As far as I can see the only discussion pending is whether the `NullableSingleInputAggregationFunction` should be merged with `BaseSingleInputAggregationFunction`

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:
```diff
@@ -263,11 +264,15 @@ public void aggregate(int length, AggregationResultHolder aggregationResultHolde
-      for (int i = 0; i < length; i++) {
-        dictIdValueMap.merge(dictIds[i], 1, Integer::sum);
-      }
+      int[] dids = blockValSet.getDictionaryIdsSV();
```
(nit) Shall we keep the name as `dictIds` and `dictId`? That is used across all the engine code

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:
```diff
@@ -263,11 +264,15 @@ public void aggregate(int length, AggregationResultHolder aggregationResultHolde
+      int[] dids = blockValSet.getDictionaryIdsSV();
+      forEachNotNull(length, blockValSet, (from, to) -> {
+        for (int i = from; i < to; i++) {
+          int did = dids[i];
```
(nit) inline

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:
```diff
@@ -0,0 +1,263 @@
+    }
+
+    @Override
+    public IntIterator clone() {
```
Do we need this override? Seems strange to me as we should never clone it

gortiz replied:
IntIterator.close is abstract in RoaringBitmaps, so we have to implement it. Anyway, I've just remove this code to introduce it in the percentille PR, which is the one that actually uses it

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:
```diff
@@ -0,0 +1,263 @@
+  /**
+   * Returns an iterator that returns the union of the null bitmaps of valSet1 and valSet2.
+   */
+  public IntIterator orNullIterator(BlockValSet valSet1, BlockValSet valSet2) {
```
Is this used? Trying to understand where this method fits

gortiz replied:
It is not used here, but it is being used in other PRs. Given I have so many PRs sharing the same base code, I've been merging them whenever I found an issue in the other PRs. For example, this method is being used in https://github.com/apache/pinot/pull/12271.

I guess I can remove it from here if I carefully merge them later

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:
```diff
@@ -467,7 +501,11 @@ public ColumnDataType getFinalResultColumnType() {
   public Double extractFinalResult(Map<? extends Number, Long> intermediateResult) {
     if (intermediateResult.isEmpty()) {
-      return DEFAULT_FINAL_RESULT;
+      if (_nullHandlingEnabled) {
```
When `_nullHandlingEnabled` is `true`, and all input values are `null`, will this map be `null`?
Same for `merge()`

gortiz replied:
_this map_ = `intermediateResult`?

In that case, no. `intermediateResult` is created in `extractAggregationResult(AggregationResultHolder)`, which calls `extractIntermediateResult` which returns not null.

gortiz replied:
I'm adding a new test to show the result when all stored data is null

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/FieldConfig.java:
```diff
@@ -254,6 +256,24 @@ public Builder withTierOverwrites(JsonNode tierOverwrites) {
       return this;
     }
 
+    public Builder withIndex(String indexId, IndexConfig indexConfig) {
```
Is this related?

gortiz replied:
I think it is the same case than `orNull` int iterator. I'm going to remove it for now

Jackie-Jiang reviewed (commented):
Mostly good. I saw you added mode into the `AllNullQueriesTest`, but somehow I didn't find where the `null` intermediate result is handled

gortiz commented:
> Mostly good. I saw you added mode into the AllNullQueriesTest, but somehow I didn't find where the null intermediate result is handled

The change I made there was to create one test per query instead of executing all in the same test method. The only reason to apply the change is to improve the UX in the case there is at least one failure. Before only the first error was reported and once you fix the issue there you had to run the tests again to check if there were other failures. Now all failures are reported at the same time.

I've modified the code long ago, so I don't remember the exact changes, but reading the diff it doesn't look I've changed anything related to null intermediate result.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ModeAggregationFunction.java:
```diff
@@ -263,11 +264,14 @@ public void aggregate(int length, AggregationResultHolder aggregationResultHolde
-      for (int i = 0; i < length; i++) {
-        dictIdValueMap.merge(dictIds[i], 1, Integer::sum);
-      }
+      int[] dicts = blockValSet.getDictionaryIdsSV();
```
lol, you really hate the name `dictIds`

Jackie-Jiang reviewed (approved):
LGTM
